rad.1
.TH RAD 1
.SH NAME
rad \- command-line calculator
.SH SYNOPSIS
.B rad
.SH DESCRIPTION
.PP
Rad reads postfix arithmetic expressions from \fIstdin\fP and prints results to
\fIstdout\fP. Input is read line by line. The top of the stack (usually the 
result of an operation) is printed to output after each newline. Each input line
is interpreted as a space-separated list of tokens. A token is either a number
or an operator. Unrecognized tokens are interpreted as the number zero.
.PP
Numbers are normally specified in decimal base, but can also be given in octal,
with prefix \fB0\fR, or in hexadecimal, with prefix \fB0x\fR (or \fB0X\fR). Some
operations are only well-defined for integer numbers, which are specified
without a radix point (\fB.\fR). Some operations accept complex numbers, which
are specified with an imaginary part after a semicolon (e.g. \fB3;4\fR). Three
numerical constants are defined: \fBpi\fR, \fBe\fR and \fBi\fR. Every number
token is stored as is into the stack.
.PP
Operators read numbers from the stack and perform computations on them, putting
the results back on the stack. For example, to perform subtraction, three input
tokens are needed, in this order: the minuend number, the subtrahend number, and
the minus sign (\fB-\fR). The minus sign is an operator that takes two numbers
from the stack, subtracts the top number from the one below it, and puts the
result back on the top of the stack.
.SH OPERATORS
.PP
The signatures inside square brackets bellow specify how many numbers each
operator takes from the stack and how many it puts back on it.
.TP
.B + - * /
Basic arithmetic operations [a b -- c]
.TP
.B %
Remainder of division (c = a-floor(a/b)*b) [a b -- c]
.TP
.B **
Exponentiation [a b -- c]
.TP
.B sqrt
Square root [a -- b]
.TP
.B ~
Bitwise not (as in C) [a -- b]
.TP
.B | & ^ >> <<
Bitwise operations (as in C) [a b -- c]
.TP
.B || && == != < <= > >=
Logic operations (as in C) [a b -- c]
.TP
.B if
If a is nonzero, return b, otherwise return c (d = a ? b : c) [a b c -- d]
.TP
.B min max
Minimum and maximum of two numbers [a b -- c]
.TP
.B log exp
Base \fBe\fR logarithm and exponentiation [a -- b]
.TP
.B floor ceil
Floor and ceiling functions [a -- b]
.TP
.B real imag
Real and imaginary part [a -- b]
.TP
.B conj
Complex conjugate [a -- b]
.TP
.B abs
Absolute value (a.k.a. modulus, magnitude) [a -- b]
.TP
.B arg
Argument (a.k.a. phase) [a -- b]
.TP
.B proj
Projection on Riemann sphere [a -- b]
.TP
.B deg rad
Angle conversion, to and from degrees [a -- b]
.TP
.B cos sin tan acos asin atan
Trigonometric functions and their inverse [a -- b]
.TP
.B atan2
Arc tangent of a/b with quadrant correction (as in C) [a b -- c]
.TP
.B cosh sinh tanh acosh asinh atanh
Hyperbolic functions and their inverse [a -- b]
.TP
.B gcd lcm
Greatest common divisor and least common multiple [a b -- c]
.TP
.B egcd
Extended GCD (ax + by = d) [a b -- x y d]
.TP
.B tot
Euler's totient function (b = |{1<=k<=a \\ gcd(k,a)==1}|) [a -- b]
.TP
.B !
Factorial function (b = product of positive integers <= a) [a -- b]
.TP
.B #
Primorial function (b = product of primes <= a) [a -- b]
.TP
.B pn
Nth prime number [n -- p]
.TP
.B isprime
1 if n is prime, 0 otherwise [n -- b]
.TP
.B prev next
Closest prime number below or above n [n -- p]
.TP
.B factor
Values and count of factors of n [n -- f1 f2 f3 ... c]
.SH STACK OPERATORS
.TP
.B drop
Drop top item from the stack [a -- ].
.TP
.B dup
Duplicate top item in the stack [a -- a a].
.TP
.B exch
Swap two items on the top of the stack [a b -- b a].
.TP
.B clear
Empty stack [a b c ... -- ]
.TP
.B :N
Copy stack item indexed by N. Indexes are 0, 1, 2, ..., -3, -2, -1.
.TP
.B show
Print stack contents.
.SH NOTES
.PP
Project homepage: <https://github.com/lecram/rad>.


rad.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <complex.h>

#define PI 3.14159265358979323846
#define E  2.71828182845904523536

#define BUFSZ 1024
static char buffer[BUFSZ];

typedef enum {DEC, HEX} repr_t;
static repr_t repr = DEC;
static int prec = 8;

typedef complex number_t;

typedef struct {
    unsigned bulk;
    unsigned len;
    number_t *slots;
} stk_t;

static stk_t stk = {16u, 0u, NULL};

static void
push(number_t number)
{
    if (stk.len >= stk.bulk) {
        stk.bulk += stk.bulk >> 1;
        stk.slots = (number_t *) realloc(stk.slots, stk.bulk * sizeof(number_t));
    }
    stk.slots[stk.len++] = number;
}

static number_t
pop() { return stk.len ? stk.slots[--stk.len] : 0; }

static number_t
peek() { return stk.len ? stk.slots[stk.len - 1] : 0; }

static void
exch()
{
    number_t n;
    n = stk.slots[stk.len - 1];
    stk.slots[stk.len - 1] = stk.slots[stk.len - 2];
    stk.slots[stk.len - 2] = n;
}

static long
gcd(long a, long b)
{
    long t;
    while (a && b) {
        t = a;
        a = b;
        b = t % b;
    }
    return a;
}

static long
lcm(long a, long b) { return a * b / gcd(a, b); }

static long
egcd(long a, long b)
{
    long q, r, x, y;
    long x1, y1, x2, y2;
    if (!(a && b)) {
        push(1);
        push(0);
    } else {
        x1 = 0; y1 = 1;
        x2 = 1; y2 = 0;
        while (b) {
            q = a / b;
            r = a % b;
            x = x2 - q * x1;
            y = y2 - q * y1;
            a = b; b = r;
            x2 = x1; x1 = x;
            y2 = y1; y1 = y;
        }
        push(x2);
        push(y2);
    }
    return a;
}

static unsigned
factorial(unsigned n)
{
    long f = 1;
    while (n) f *= n--;
    return f;
}

static unsigned
tot(unsigned n)
{
    unsigned tot = 1;
    unsigned pos = n - 1;
    while (pos > 1) {
        if (gcd((long) pos, (long) n) == 1)
            tot++;
        pos--;
    }
    return tot;
}

static unsigned
isprime(unsigned n)
{
    unsigned d;
    unsigned m;
    if (n < 2)
        return 0;
    if (n == 2)
        return 1;
    if (n % 2 == 0)
        return 0;
    m = (unsigned) sqrt((double) n);
    for (d = 3; d <= m; d += 2)
        if (n % d == 0)
            return 0;
    return 1;
}

static unsigned
prev(unsigned n)
{
    if (n < 4)
        return 2;
    if (n % 2 == 0)
        n++;
    for (n -= 2; n > 2; n -= 2)
        if (isprime(n))
            break;
    return n;
}

static unsigned
next(unsigned n)
{
    if (n < 2)
        return 2;
    if (n == 2)
        return 3;
    if (n % 2 == 0)
        n--;
    for (n += 2; n > 0; n += 2)
        if (isprime(n))
            break;
    return n;
}

static unsigned
pn(unsigned n)
{
    unsigned p = 1;
    while (n--) p = next(p);
    return p;
}

static unsigned
primorial(unsigned n)
{
    unsigned r;
    r = isprime(n) ? n : 1;
    while (n > 2) {
        n = prev(n);
        r *= n;
    }
    return r;
}

static unsigned
factor(unsigned n)
{
    unsigned d;
    unsigned c = 0;
    if (n % 2 == 0) {
        push(2);
        c++;
        while (n % 2 == 0) n /= 2;
    }
    for (d = 3; d <= n; d += 2) {
        if (n % d == 0) {
            push(d);
            c++;
            while (n % d == 0) n /= d;
        }
    }
    return c;
}

static int
snprintn(char *buf, int buf_size, number_t number)
{
    int count = 0;
    switch (repr) {
        case DEC:
            if (cimag(number))
                count = snprintf(
                    buf, buf_size, "%.*g;%.*g",
                    prec, creal(number), prec, cimag(number)
                );
            else
                count = snprintf(
                    buf, buf_size, "%.*g",
                    prec, creal(number)
                );
            break;
        case HEX:
            count = snprintf(buf, buf_size, "%#0*X", sizeof(unsigned) * 2 + 2, (unsigned) number);
            break;
    }
    return count;
}

static number_t
parse(const char *token)
{
    number_t number;
    if (!strcmp(token, "e"))
        return E;
    else if (!strcmp(token, "pi"))
        return PI;
    else if (!strcmp(token, "i"))
        return I;
    else if (strchr(token, ';')) {
        char *sep;
        number = strtod(token, &sep);
        number += atof(++sep) * I;
    } else if (strchr(token, '.'))
        number = (number_t) atof(token);
    else
        number = (number_t) strtol(token, NULL, 0);
    return number;
}

#define BINOP(OP)   else if (!strcmp(token, #OP)) \
                    { b = pop(); a = pop(); push(a OP b); }

#define BINBOP(OP)  else if (!strcmp(token, #OP)) \
                    { b = pop(); a = pop(); push(((unsigned) a) OP ((unsigned) b)); }

#define BINROP(OP)   else if (!strcmp(token, #OP)) \
                    { b = pop(); a = pop(); push(creal(a) OP creal(b)); }

#define BFUNC(OP)   else if (!strcmp(token, #OP)) \
                    { b = pop(); a = pop(); push(OP((long) a, (long) b)); }

#define CFUNC(OP)   else if (!strcmp(token, #OP)) \
                    { push(c ## OP(pop())); }

#define UFUNC(OP)   else if (!strcmp(token, #OP)) \
                    { push(OP((unsigned) pop())); }

static void
process(const char *token)
{
    number_t a, b;
    if (!strcmp(token, "prec"))
        prec = pop();
    BINOP(+) BINOP(-) BINOP(*) BINOP(/)
    BINOP(==) BINOP(!=) BINOP(||) BINOP(&&)
    BINBOP(|) BINBOP(&) BINBOP(^) BINBOP(>>) BINBOP(<<)
    BINROP(<) BINROP(<=) BINROP(>) BINROP(>=)
    BFUNC(gcd) BFUNC(lcm) BFUNC(egcd)
    CFUNC(real) CFUNC(imag) CFUNC(arg) CFUNC(abs) CFUNC(proj) CFUNC(exp)
    CFUNC(log) CFUNC(sqrt) CFUNC(acos) CFUNC(asin) CFUNC(atan) CFUNC(cos)
    CFUNC(sin) CFUNC(tan) CFUNC(acosh) CFUNC(asinh) CFUNC(atanh) CFUNC(cosh)
    CFUNC(sinh) CFUNC(tanh)
    UFUNC(tot) UFUNC(factor) UFUNC(isprime) UFUNC(prev) UFUNC(next) UFUNC(pn)
    else if (!strcmp(token, "%")) {
        b = pop(); a = pop();
        push(fmod(creal(a), creal(b)));
    } else if (!strcmp(token, "**")) {
        b = pop(); a = pop();
        push(cpow(a, b));
    } else if (!strcmp(token, "~")) {
        push(~ (unsigned) pop());
    } else if (!strcmp(token, "floor")) {
        push(floor(pop()));
    } else if (!strcmp(token, "ceil")) {
        push(ceil(pop()));
    } else if (!strcmp(token, "deg")) {
        push(creal(pop()) * 180 / PI);
    } else if (!strcmp(token, "rad")) {
        push(creal(pop()) * PI / 180);
    } else if (!strcmp(token, "atan2")) {
        b = pop(); a = pop();
        push(atan2(creal(a), creal(b)));
    } else if (!strcmp(token, "conj")) {
        push(conj(pop()));
    } else if (!strcmp(token, "!")) {
        push(factorial((unsigned) pop()));
    } else if (!strcmp(token, "#")) {
        push(primorial((unsigned) pop()));
    } else if (!strcmp(token, "if")) {
        b = pop(); a = pop();
        push(pop() ? a : b);
    } else if (!strcmp(token, "min")) {
        b = pop(); a = pop();
        push(creal(a) < creal(b) ? a : b);
    } else if (!strcmp(token, "max")) {
        b = pop(); a = pop();
        push(creal(a) > creal(b) ? a : b);
    } else if (!strcmp(token, "drop")) {
        pop();
    } else if (!strcmp(token, "dup")) {
        push(peek());
    } else if (!strcmp(token, "exch")) {
        exch();
    } else if (!strcmp(token, "clear")) {
        while (stk.len) pop();
    } else if (token[0] == ':') {
        int index = strtol(token+1, NULL, 0);
        if (index < 0) index += stk.len;
        push(index >= 0 && index < stk.len ? stk.slots[index] : 0);
    } else
        push(parse(token));
}

int
main()
{
    const char *token;
    stk.slots = (number_t *) malloc(stk.bulk * sizeof(number_t));
    while (!feof(stdin)) {
        if (!fgets(buffer, BUFSZ, stdin))
            break;
        if (!strcmp(buffer, "show\n")) {
            unsigned i;
            int offset;
            for (i = 0, offset = 0; i < stk.len; i++, offset++) {
                offset += snprintn(buffer + offset, BUFSZ - offset, stk.slots[i]);
                strcat(buffer, " ");
            }
            puts(buffer);
        } else if (!strcmp(buffer, "dec\n")) {
            repr = DEC;
        } else if (!strcmp(buffer, "hex\n")) {
            repr = HEX;
        } else {
            token = strtok(buffer, " \n");
            while (token) {
                process(token);
                token = strtok(NULL, " \n");
            }
            if (stk.len) {
                snprintn(buffer, BUFSZ, peek());
                puts(buffer);
            }
        }
    }
    free(stk.slots);
    return 0;
}
